frontend/src/App.vue

```diff
@@ -220,30 +221,69 @@ img {
 <script lang="ts">
 import { defineComponent } from "vue";
 
-// State providers
-import dialog from "@/utilities/dialog";
-import documents from "@/utilities/documents";
-import fullscreen from "@/utilities/fullscreen";
+import { DialogState, createDialogState } from "@/state/dialog";
+import { createDocumentsState, DocumentsState } from "@/state/documents";
+import { createFullscreenState, FullscreenState } from "@/state/fullscreen";
 
 import MainWindow from "@/components/window/MainWindow.vue";
 import LayoutRow from "@/components/layout/LayoutRow.vue";
+import { createEditorState, EditorState } from "@/state/wasm-loader";
+import { createInputManager, InputManager } from "@/lifetime/input";
+import { initErrorHandling } from "@/lifetime/errors";
+
+// Vue injects don't play well with TypeScript, and all injects will show up as `any`. As a workaround, we can define these types.
+declare module "@vue/runtime-core" {
+	interface ComponentCustomProperties {
+		dialog: DialogState;
+		documents: DocumentsState;
+		fullscreen: FullscreenState;
+		editor: EditorState;
+		inputManger?: InputManager;
```

Comment:
Why is this optional? Please comment to avoid confusion (or fix it if it's better being not an optional parameter).

Reply:
Because the inputManager is not initialized in `data()` (because the div hasn't been created yet), but in `mounted()`. So there is a time in the lifecycle of the component where `inputManager = undefined`, and the type must reflect that.

Reply:
Gotcha, I'll add that comment now then, thanks for explaining.